python/cudf/cudf/core/series.py

```diff
@@ -216,17 +216,25 @@ def __setitem__(self, key, value):
         if (
             not isinstance(
                 self._frame._column.dtype,
-                (cudf.core.dtypes.DecimalDtype, cudf.CategoricalDtype),
+                (
+                    # Casting only for string and non-decimal numeric
```

Comment:
Can we change this to an “allow list” by using the same `_is_non_decimal_numeric_dtype` as we use for `value`? Something like “is (non-decimal) numerical or string” rather than “not decimal, categorical, list, or struct”.

Also it feels like there should be a clearer and more perfect symmetry in the value and column dtypes we allow here. 🤔 Does pandas only allow strings for the Series but not the scalar value, or something like that?

Reply:
> I think I’m asking a different question. This code appears to permit a string column to be assigned an int value (a string column is not disallowed by the excluded dtypes for decimal, categorical, list, struct). However, the opposite seems to not be allowed because the value must be “non-decimal numeric.” Is that sensible? Why this asymmetry? Can those paths be added to tests?

stringcolumn gets int value (say) is "fine" because we can do `v.astype(column.dtype)` to get a `cudf.Scalar` with `dtype('O')` and then `libcudf::scatter` works fine because we the value we are scattering is not `10` but `"10"`.

One can assign a string to an int column iff `to_cudf_compatible_scalar(some_str, int)` is successful (the resulting column is still of dtype int, rather than upcasting to dtype str).

AKA, this is all very inconsistent.

Reply:
Oh wow. Okay, I assumed the latter would force all values to strings. I trust this more now but would like tests for those.

Reply:
Added something.

Reply:
@wence- Is it possible to tackle the first part of my suggestion? Or did I miss something in the thread about why that isn't a good idea?

> Can we change this to an “allow list” by using the same _is_non_decimal_numeric_dtype as we use for value? Something like “is (non-decimal) numerical or string” rather than “not decimal, categorical, list, or struct”.

Reply:
No good reason. I think I've done that now.